src/librustc_typeck/check/mod.rs

```diff
@@ -3222,6 +3222,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             formal_tys.clone()
         };
 
+        let mut final_arg_types: Vec<(usize, Ty<'_>)> = vec![];
```

Comment:
Seems like this is going to regress happy-path compile-time perf? (I don't expect by much but still...)

Reply:
I see no way of doing this without collecting the resolved types for the general case (there's a lazy way to do it that would only work with fully resolved types in the arguments).

Reply:
Yeah sure, just pointing out that there is a cost to this, which may be totally acceptable. :)